compiler/p/codegen/ControlFlowEvaluator.cpp

```diff
@@ -1693,864 +2360,1029 @@ TR::Register *OMR::Power::TreeEvaluator::ifacmpneEvaluator(TR::Node *node, TR::C
+ *
+ * \param compareInfo
+ *    The CompareInfo struct containing information about the comparison.
+ *
+ * \param cg
+ *    The code generator.
+ *
+ * \return
+ *    true if the comparison is a signed comparison, does not use register pairs, and is against
+ *    a value with a known sign bit; false otherwise.
+ */
+bool isSimpleSignedCompareToKnownSign(TR::Node *secondChild, const CompareInfo& compareInfo, TR::CodeGenerator *cg)
+   {
+   if (compareInfo.isUnsigned)
+      return false;
 
-   node->setRegister(trgReg);
-   cg->decReferenceCount(firstChild);
-   cg->decReferenceCount(secondChild);
-   return trgReg;
+   if (compareInfo.type == TR::Int64 && !cg->comp()->target().is64Bit())
+      return false;
+
+   return secondChild->isNonNegative() || secondChild->isNegative() || secondChild->getOpCode().isLoadConst();
    }
 
-void OMR::Power::TreeEvaluator::genBranchSequence(TR::Node* node,
-   TR::Register* src1Reg, TR::Register* src2Reg, TR::Register* trgReg,
-   TR::InstOpCode::Mnemonic branchOpCode, TR::InstOpCode::Mnemonic cmpOpCode, TR::CodeGenerator* cg)
+/**
+ * \brief
+ *    Gets the length of an integral type in bits.
+ *
+ * \param dt
+ *    The data type to get the length of.
+ *
+ * \param cg
+ *    The code generator.
+ *
+ * \return
+ *    The length of the provided type in bits.
+ */
+int32_t getTypeBitLength(TR::DataTypes dt, TR::CodeGenerator *cg)
    {
-   TR::Register    *condReg = cg->allocateRegister(TR_CCR);
-   TR::PPCControlFlowInstruction *cfop = (TR::PPCControlFlowInstruction *)
-      generateControlFlowInstruction(cg, TR::InstOpCode::setbool, node);
-   cfop->addTargetRegister(condReg);
-   cfop->addTargetRegister(trgReg);
-   cfop->addSourceRegister(src1Reg);
-   cfop->addSourceRegister(src2Reg);
-   cfop->setOpCode2Value(branchOpCode);
-   cfop->setCmpOpValue(cmpOpCode);
-   cg->stopUsingRegister(condReg);
+   switch (dt)
+      {
+      case TR::Int8:
+         return 8;
+      case TR::Int16:
+         return 16;
+      case TR::Int32:
+         return 32;
+      case TR::Int64:
+         return 64;
+      case TR::Address:
+         return cg->comp()->target().is64Bit() ? 64 : 32;
+      default:
+         TR_ASSERT_FATAL(false, "Invalid data type %s for getTypeBitLength", TR::DataType::getName(dt));
+      }
    }
 
+/**
+ * \brief
+ *    Determines whether an integral type is smaller than the size of a GPR on the current
+ *    machine.
+ *
+ * \param dt
+ *    The data type to check.
+ *
+ * \param cg
+ *    The code generator.
+ *
+ * \return
+ *    true if the provided type is smaller than the size of a GPR; false otherwise.
+ */
+bool isTypeSubRegister(TR::DataTypes dt, TR::CodeGenerator *cg)
+   {
+   switch (dt)
+      {
+      case TR::Int8:
+      case TR::Int16:
+         return true;
+      case TR::Int32:
+         return cg->comp()->target().is64Bit();
+      case TR::Int64:
+      case TR::Address:
+         return false;
```

Comment:
target().is32Bit() only means that JVM is running in 32bit mode of the hardware.  It doesn't mean the machine is not 64-bit capable or its GPR is 32bit. TR::Address could be a sub-register type in 32bit JVM mode. And, actually it must be, since we switched to support 64bit CPU/hardware only.

On POWER though, depending on where you will use isTypeSubRegitser() result, false on TR::Address could be harmless (because there are cmp8 and cmp4 instructions to make it).  If it is used for sign-bit decision, it could be a bug.  I have impression of checking this code before.  Did you change this from previous version?

Reply:
even if this could be harmless, better fix it to the truth.

Reply:
This function is meant to be used to check if there are unused high bits in the register that could be used as part of an optimization. In this case, this is guarding an optimization that can compare two values by sign- or zero-extending them and then subtracting them and checking the sign bit of the result, which is only possible when the result cannot overflow. The fact that the hardware technically has extra bits in the register is irrelevant to whether or not we're meant to be using them as an optimization. I've added some text to the Doxygen comment to clarify this.